File: python/atomic_kernels/viewer/_camera.py
from __future__ import annotations

from typing import Iterable


def _structure_to_world(vector: Iterable[float]) -> tuple[float, float, float]:
    x, y, z = (float(component) for component in vector)
    return (x, z, -y)


class CameraController:
    """Control the viewer camera in structure-space coordinates.

    The viewer renders chemistry-style structure coordinates with ``z`` as the
    semantic up axis. Internally the Bevy camera still uses its own world
    convention, so focus points and pan vectors are rotated before being sent to
    the low-level backend.
    """

    def __init__(self, session: "ViewerSessionFacade") -> None:
        self._session = session

    def look_at(
        self,
        focus,
        radius: float | None = None,
        yaw: float | None = None,
        pitch: float | None = None,
    ) -> None:
        """Set the camera focus and optionally update view parameters.

        Parameters
        ----------
        focus
            Structure-space ``(x, y, z)`` focus point.
        radius : float or None, optional
            Camera distance from the focus point.
        yaw : float or None, optional
            Orbit angle around the structure ``z`` axis, in radians.
        pitch : float or None, optional
            Tilt away from the structure ``xy`` plane, in radians.
        """
        self._session._backend.set_camera_view(
            _structure_to_world(focus), radius, yaw, pitch
        )

    def set_focus(self, focus) -> None:
        """Move the camera focus to a new structure-space point.

        Parameters
        ----------
        focus
            Structure-space ``(x, y, z)`` focus point.
        """
        self._session._backend.set_camera_view(
            _structure_to_world(focus), None, None, None
        )

    def set_rotation(
        self, yaw: float | None = None, pitch: float | None = None
    ) -> None:
        """Set the camera yaw and pitch in structure-space terms.

        Parameters
        ----------
        yaw : float or None, optional
            Orbit angle around the structure ``z`` axis, in radians.
        pitch : float or None, optional
            Tilt away from the structure ``xy`` plane around structure ``x``, in radians.
        """
        self._session._backend.set_camera_view(None, None, yaw, pitch)

    def set_radius(self, radius: float) -> None:
        """Set the camera distance from the current focus point.

        Parameters
        ----------
        radius : float
            Camera distance from the current focus point.
        """
        self._session._backend.set_camera_view(None, radius, None, None)

    def pan(self, delta) -> None:
        """Translate the camera focus by a structure-space delta.

        Parameters
        ----------
        delta
            Structure-space ``(x, y, z)`` translation applied to the focus point.
        """
        self._session._backend.pan_camera(_structure_to_world(delta))

    def zoom(
        self, factor: float | None = None, delta: float | None = None
    ) -> None:
        """Zoom by a multiplicative factor or additive radius delta.

        Parameters
        ----------
        factor : float or None, optional
            Multiplicative zoom factor.
        delta : float or None, optional
            Additive change in camera radius.
        """
        if (factor is None) == (delta is None):
            raise ValueError("zoom requires exactly one of factor or delta")
        self._session._backend.zoom_camera(factor=factor, delta=delta)

    def orbit(self, yaw_delta: float = 0.0, pitch_delta: float = 0.0) -> None:
        """Apply an immediate orbit step.

        Parameters
        ----------
        yaw_delta : float, default=0.0
            Increment around the structure ``z`` axis, in radians.
        pitch_delta : float, default=0.0
            Increment away from the structure ``xy`` plane, in radians.
        """
        self._session._backend.orbit_camera(yaw_delta, pitch_delta)

    def frame_all(self) -> None:
        """Reset the camera to frame the whole current structure."""
        self._session._backend.frame_all()

    def start_orbit(self, yaw_rate: float = 0.5, pitch_rate: float = 0.0) -> None:
        """Start continuous orbit motion.

        Parameters
        ----------
        yaw_rate : float, default=0.5
            Continuous rotation rate around the structure ``z`` axis, in radians per second.
        pitch_rate : float, default=0.0
            Continuous tilt rate, in radians per second.
        """
        self._session._backend.start_orbit(yaw_rate, pitch_rate)

    def stop(self) -> None:
        """Stop any active camera motion started by ``start_orbit``."""
        self._session._backend.stop_camera_motion()

File: crates/ak-vis/src/viewer/session/tests/camera.rs
use super::helpers::*;
use crate::viewer::session::{CameraState, ViewerCommand, ViewerState, camera_view_for_frame};
use ak_core::Trajectory;
use bevy::prelude::Vec3;
#[test]
fn set_camera_view_updates_requested_fields() {
    let viewer = ViewerState::new(Trajectory::new(vec![test_structure(0.0)]), 0);
    let mut camera = CameraState::new(&viewer);

    camera.apply_command(
        &viewer,
        &ViewerCommand::SetCameraView {
            focus: Some([1.0, 2.0, 3.0]),
            radius: Some(9.0),
            yaw: Some(0.5),
            pitch: Some(-0.25),
        },
    );

    assert_eq!(camera.focus, Vec3::new(1.0, 2.0, 3.0));
    assert_eq!(camera.radius, 9.0);
    assert_eq!(camera.yaw, 0.5);
    assert_eq!(camera.pitch, -0.25);
}

#[test]
fn orbit_and_zoom_camera_are_incremental() {
    let viewer = ViewerState::new(Trajectory::new(vec![test_structure(0.0)]), 0);
    let mut camera = CameraState::new(&viewer);
    let initial = camera.clone();

    camera.apply_command(
        &viewer,
        &ViewerCommand::OrbitCamera {
            yaw_delta: 0.2,
            pitch_delta: -0.1,
        },
    );
    camera.apply_command(
        &viewer,
        &ViewerCommand::ZoomCamera {
            factor: Some(0.5),
            delta: None,
        },
    );

    assert_eq!(camera.yaw, initial.yaw + 0.2);
    assert_eq!(camera.pitch, initial.pitch - 0.1);
    assert_eq!(camera.radius, initial.radius * 0.5);
}

#[test]
fn start_and_stop_orbit_motion_updates_camera() {
    let viewer = ViewerState::new(Trajectory::new(vec![test_structure(0.0)]), 0);
    let mut camera = CameraState::new(&viewer);
    let initial_yaw = camera.yaw;

    camera.apply_command(
        &viewer,
        &ViewerCommand::StartOrbit {
            yaw_rate: 1.0,
            pitch_rate: 0.0,
        },
    );
    camera.tick_motion(0.5);
    assert_eq!(camera.yaw, initial_yaw + 0.5);

    camera.apply_command(&viewer, &ViewerCommand::StopCameraMotion);
    camera.tick_motion(0.5);
    assert_eq!(camera.yaw, initial_yaw + 0.5);
}

#[test]
fn frame_all_restores_default_camera_view() {
    let viewer = ViewerState::new(Trajectory::new(vec![test_structure(0.0)]), 0);
    let mut camera = CameraState::new(&viewer);
    camera.focus = Vec3::splat(5.0);
    camera.radius = 99.0;
    camera.yaw = 2.0;
    camera.pitch = 1.0;

    camera.apply_command(&viewer, &ViewerCommand::FrameAll);

    let expected = camera_view_for_frame(&viewer).unwrap();
    assert_eq!(camera.focus, expected.focus);
    assert_eq!(camera.radius, expected.radius);
    assert_eq!(camera.yaw, expected.yaw);
    assert_eq!(camera.pitch, expected.pitch);
}
